arangod/RestHandler/RestReplicationHandler.cpp

```diff
@@ -2893,6 +2920,319 @@ void RestReplicationHandler::handleCommandLoggerTickRanges() {
+                  "could not get replication iterator for collection '" +
+                      ctx.cname + "'");
+    return false;
+  }
+
+  return true;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief return the revision tree for a given collection, if available
+/// @response serialized revision tree, binary
+//////////////////////////////////////////////////////////////////////////////
+
+void RestReplicationHandler::handleCommandRevisionTree() {
+  RevisionOperationContext ctx;
+  if (!prepareRevisionOperation(ctx)) {
+    return;
+  }
+
+  auto tree = ctx.collection->getPhysical()->revisionTree(ctx.batchId);
+  if (!tree) {
+    generateError(rest::ResponseCode::SERVER_ERROR, TRI_ERROR_INTERNAL,
+                  "could not generate revision tree");
+    return;
+  }
+  std::string result = tree->serialize();
+
+  resetResponse(rest::ResponseCode::OK);
+  _response->setContentType("application/octet-stream");
+  _response->addRawPayload(VPackStringRef(result.data(), result.size()));
+}
+
+void RestReplicationHandler::handleCommandRebuildRevisionTree() {
+  RevisionOperationContext ctx;
+  if (!prepareRevisionOperation(ctx)) {
+    return;
+  }
+
+  if (!ctx.collection->syncByRevision()) {
+    generateError(Result(TRI_ERROR_BAD_PARAMETER,
+                         "collection does not support revision tree commands"));
+    return;
+  }
+
+  Result res = ctx.collection->getPhysical()->rebuildRevisionTree();
+  if (res.fail()) {
+    generateError(res);
+    return;
+  }
+
+  generateResult(rest::ResponseCode::NO_CONTENT, VPackSlice::nullSlice());
+}
+
+//////////////////////////////////////////////////////////////////////////////
+/// @brief return the requested revision ranges for a given collection, if
+///        available
+/// @response VPackObject, containing
+///           * ranges, VPackArray of VPackArray of revisions
+///           * resume, optional, if response is chunked; revision resume
+///                     point to specify on subsequent requests
+//////////////////////////////////////////////////////////////////////////////
+
+void RestReplicationHandler::handleCommandRevisionRanges() {
+  RevisionOperationContext ctx;
+  if (!prepareRevisionOperation(ctx)) {
+    return;
+  }
+
+  bool success = false;
+  std::size_t current = 0;
+  VPackSlice const body = this->parseVPackBody(success);
+  if (!success) {
+    // error already created
+    return;
+  }
+  bool badFormat = !body.isArray();
+  if (!badFormat) {
+    for (std::size_t i = 0; i < body.length(); ++i) {
+      if (!body.at(i).isArray() || body.at(i).length() != 2 ||
+          !body.at(i).at(0).isNumber() || !body.at(i).at(1).isNumber()) {
+        badFormat = true;
+        break;
+      }
+      if (body.at(i).at(0).getNumber<std::size_t>() <= ctx.resume ||
+          (i > 0 && body.at(i - 1).at(1).getNumber<std::size_t>() < ctx.resume)) {
+        current = std::max(current, i);
+      }
```

Comment:
Repeated invocations of `velocypack::Slice::at(...)` may be expensive.
`at` is stateless, and each invocation needs to figure out the size of the underlying velocypack Array, where the index table is etc.
Using a `velocypack::ArrayIterator` will solve these problems.